### DISCUSSION TITLE
Improving typing custom key extractors for urql-graphcache-exchange

### DISCUSSION THREAD

**Discussion by jedahan** [urql-graphql/urql#2172] [Ideas]
I have a field `artistId: string` which acts as the id for one of my graphql types.

The cacheExchange `KeyingConfig` type doesn't know what fields I have, so we need to tell typescript. Maybe we can make it a `Record<any, string>` or make a generic `KeyExtractField<T: SomeTypeScriptType>` to allow passing in a known compile-time type?

Fails with "string | null" -> number type error:

```ts
cacheExchange({
  keys: {
    Artist: (artist) => artist.artistId,
  },
})
```

Works fine

```ts
cacheExchange({
  keys: {
    Artist: (artist: Data & { artistId: string }) =>
      artist.artistId,
    },
  }
})
```

**Comment by JoviDeCroock** [comment]
I think something like this should be able to help https://github.com/FormidableLabs/urql/issues/901#issuecomment-840521982

**Reply by jedahan** [reply]
Excellent, will explore shortly
